feat: New™️ easy-install.py

athul opened the pull request:
This PR includes
- New containerized setup of Frappe and ERPNext
-  Uses frappe_docker
- Generates compose files on the fly and spins up a new production server
- Installs and sets up docker if not present
- Made with standard library
- Can be used for development and production setup

Special thanks to @revant for helping me out on this <3


closes https://github.com/frappe/frappe/issues/18223

---

TODO:
- [x] Resolve open issues
- [x] Manual tests @ankush / @revant 
- [x] Automated test for script on github action @athul 
- [x] Add "copy and run in shell" 2-3 liners in README.md of bench and frappe

athul committed:
feat: New easy-install.py

Supports docker container setup with production and dev instances

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+CRED = "\033[31m"
+CEND = "\033[0m"
+CGRN = "\33[92m"
+CYLW = "\33[93m"
```
Add function wrappers around these so we dont forget to add CEND :)

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+def clone_frappe_docker_repo() -> None:
+    try:
+        subprocess.run(
+            ["git", "clone", "https://github.com/frappe/frappe_docker"], check=True
```
This assumes git is installed, which by default isn't on most distros.

ankush replied:
This should be either part of instruction OR handled by script.

Alternatively, download the zip over HTTP and expand. We don't really need a git repository.

revant replied:
> From "broad goal" POV we should also limit ourselves to few Linux distros. Ubutnu/Debian dervatives at first and maybe Fedora someday later. We can assume that other crazy distro users can setup frappe_docker themselves.

Once we have docker installed, distro doesn't matter.
If docker is not installed, only ubuntu/debian/fedora/centos/rhel or distros supported by `curl -fsSL get.docker.com | bash` will install docker. Any unsupported distro by the script shows message like:

```
ERROR: Unsupported distribution 'alpine'
```

more about script here https://github.com/docker/docker-install

In case of unsupported distro, if someone manages to install `docker` and `docker compose` then the script will proceed.

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+    with open(os.path.join(wd, ".env"), "w") as f:
+        f.writelines(
+            [
+                "FRAPPE_VERSION=v14.17.1\n",
+                "ERPNEXT_VERSION=v14.9.0\n",
```
ummm hardcoding versions?

ankush replied:
Instead of hardcoding tags we can hard-code branches which don't change frequently. 

There's `v14` tag on docker hub which points to latest images in v14.x.x

https://hub.docker.com/layers/frappe/erpnext-worker/v14/images/sha256-500eb8a09edb45a698b42c1b42eb42441a7e8c1e44ac08d4c0746a72cd4526e5?context=explore

revant replied:
These are already part of frappe_docker/example.env, and we have a release action bumping `example.env` on every frappe or erpnext release. Just copying `example.env` to `.env` can get the latest version without hardcoding.

read/update `.env` file after you copy from `example.env`

athul replied:
If the version is provided as without minor version numbers, the latest image will be fetched. Say if I provide `FRAPPE_VERSION=v14`, the fetched image will be v14.17.1. Which is actually the latest version of Frappe rn(04/12/22)

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+        )
+
+def generate_pass(length:int=9) -> str:
+    return b64encode(token_bytes(length)).decode()
```
Note: `token_bytes` will give 2x length for specified length. If not intentional copy `generate_hash` from frappe to get exact length.

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+            ["git", "clone", "https://github.com/frappe/frappe_docker"], check=True
+        )
+    except Exception as e:
+        print(f"\n{CRED}Cloning frappe_docker Failed{CEND}\n\n", e)
```
Log entire exception somewhere so we can ask for more info when required. Printing exception only prints final message. 


This applies everywhere in the script where exceptions are caught and generic message is printed. We'd need some input from those failures to fix them :)

ankush replied:
use `logging` module with `exc_info=True`

Log as much as you can so we get decent input for fixing issues when they inevitably come.

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+
+
+def write_to_env(wd, site,db_pass,email):
+    site_name = site if site != "" else ""
```
```suggestion
    site_name = site or ""
```

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+        setup_dev_instance(args.project)
+    elif args.prod:
+        print(f"\n{CGRN}Setting Up Production Instance{CEND}\n")
+        if "example.com" in args.email:
```
why this specific check? Many dummy emails out there.

revant replied:
traefik will throw error if it is example.com for letsencrypt/certbot notification email.

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+            )
+            exit(1)
+        try:
+            ps = subprocess.run(
+                ["curl", "-fsSL", "https://get.docker.com"],
+                capture_output=True,
+                check=True,
+            )
+            subprocess.run(["/bin/bash"], input=ps.stdout, capture_output=True)
```
Run the first process as shell directly instead of manual redirection to shell?

athul replied:
Tried, didn't work 😓

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+            print(
+                f"\n\n {CYLW}Try Installing Docker Manually and re-run this script again{CEND}\n\n"
+            )
+            exit(1)
```
Use `sys.exit` 

`exit` global is for interactive scripts.

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+                f"\n\n {CYLW}Try Installing Docker Manually and re-run this script again{CEND}\n\n"
+            )
+            exit(1)
+    else:
+        return
```
```suggestion
```

This doesn't do anything

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,239 @@
+
+
+def install_docker():
+    if which("docker") is None:
```
Use guard clause to avoid over-indentation. `if which("docker") is not None: return`


https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

ankush commented:
TODO:
- [ ] Resolve open issues
- [ ] Manual tests @ankush 
- [ ] Add "copy and run in shell" 2-3 liners in README.md of bench and frappe
- [ ] Add non-erpnext install?

ankush commented:
From "broad goal" POV we should also limit ourselves to few Linux distros. Ubutnu/Debian dervatives at first and maybe Fedora someday later. We can assume that other crazy distro users can setup frappe_docker themselves.

Easy install became a mess cause it was trying to handle too many things poorly.

revant commented:
> From "broad goal" POV we should also limit ourselves to few Linux distros. Ubutnu/Debian dervatives at first and maybe Fedora someday later. We can assume that other crazy distro users can setup frappe_docker themselves.

Once we have docker installed, distro doesn't matter.
If docker is not installed, only ubuntu/debian/fedora/centos/rhel or distros supported by `curl -fsSL get.docker.com | bash` will install docker. Any unsupported distro by the script shows message like:

```
ERROR: Unsupported distribution 'alpine'
```

more about script here https://github.com/docker/docker-install

In case of unsupported distro, if someone manages to install `docker` and `docker compose` then the script will proceed.

athul committed:
chore: Added logging and removed git dependency

Also added
- New methods for printing
- Better env file management
- Better password generation
- Added zip repository downloads
- default container version set to v14

athul commented:
Worked on all the suggestions and reviews, Tested out the script multiple times on a docker container inside a container(dind). You can too

```sh
# Download the easy install script to a folder called scripts
$ docker run --rm -it -v <path-to-scripts-folder>:/root/scripts -v /var/run/docker.sock:/var/run/docker.sock python:3-alpine sh
# Once inside the container
$ apk add docker-cli docker-cli-compose curl
$ cd ~/scripts
$ python easy-install.py --prod --project easyinstall -s easy.local --email <some-email>
```

revant commented:
I tested the script.

It installs with `v14` tag, That may pull latest v14 images causing unintended non-major upgrades. I'd recommend either get the version from frappe/erpnext releases or from example.env.

athul committed:
chore: get exact versions from example.env file

revant commented:
Everything works as expected. Site is up and running on `https://easy.localhost`, http redirects to https.
easy-install.log, easyinstall-compose.yml passwords.txt and frappe_docker directory found in path from where `easy-install.py` was executed.

add cronjob if possible https://github.com/frappe/frappe_docker/blob/main/docs/backup-and-push-cronjob.md

ankush commented on easy-install.py:
```diff
@@ -0,0 +1,306 @@
+
+
+def generate_pass(length: int = 12) -> str:
+    return sha224(repr(time.time()).encode()).hexdigest()[:length]
```
time.time() isn't random enough. Copy the new version from develop branch :)

ankush replied:
```python
def generate_pass(length: int = 12) -> str:
	"""Generate random hash using best available randomness source."""
	import math
	import secrets

	if not length:
		length = 56

	return secrets.token_hex(math.ceil(length / 2))[:length]
```

athul committed:
chore: Update Readme and changed password generated method

Co-authored-by: @ankush

athul committed:
feat: add GHA tests
chore: added reading from env when rerunning script
- Makes current site as default

revant commented:
Idea for cron yml template.

```yaml
# backup-job.yml.tmpl
services:
  backup:
    image: frappe/frappe-worker:${FRAPPE_VERSION:?No Frappe version set}
    volumes:
      - sites:/home/frappe/frappe-bench/sites
      - assets:/home/frappe/frappe-bench/sites/assets:ro
    entrypoint: ["bash", "-c"]
    command: |
      for SITE in $(/home/frappe/frappe-bench/env/bin/python -c "import frappe;print(' '.join(frappe.utils.get_sites()))")
      do
        bench --site $SITE backup --with-files
      done
    volumes:
      - sites:/home/frappe/frappe-bench/sites
    networks:
      - bench-network

networks:
  bench-network:
    external: true
    name: ${project}_default

volumes:
  sites:
    external: true
    name: ${project}_sites
```

`volumes` and `networks` need to be dynamic. How about making it as a template string python code? or read from template file from frappe_docker somewhere?

For upgrade?

- frappe_docker/.env needs to update the env vars or it'll just pick up existing ones from .env
- populate assets, the acrobatics of clearing volume and populating it so that python code can read js static files. Needed for `website_theme.py`
- migrate sites, `bench --site {site-name} migrate` to be executed for site(s) available in the frappe-bench?

more about populate assets and migrate sites here https://github.com/frappe/frappe_docker/issues/1020#issuecomment-1352626691

athul commented:
@revant Will add the cron in a subsequent PR. Thanks for the inputs 💟

ankush committed:
ci: add concurrency group

ankush reviewed (approved):
Works flawlessly!

Awesome work @athul and thanks a lot for everything you do in Frappe ecosystem @revant 


:rocket:
